File: apps/agent/src/agents/planning.ts
import { ChatPromptTemplate, MessagesPlaceholder } from '@langchain/core/prompts';
import { RunnableSequence } from '@langchain/core/runnables';
import { AgentType, ActionItem } from '../types';
import { createLLM, getSystemPrompt } from '../utils/llm';
import { StructuredOutputParser } from 'langchain/output_parsers';
import { z } from 'zod';
import { AIMessage } from '@langchain/core/messages';
import { StateAnnotation } from '../types';
import { getMCPClient } from '../services/mcp-client';
import { generateBulkSubtaskCreationCode } from '../utils/code-generator';

// Process the user input with Planning agent
export const processPlanning = async (state: typeof StateAnnotation.State): Promise<typeof StateAnnotation.State> => {
  // Create LLM
  const llm = createLLM('gpt-4o', 0.2);

  // Create parser for structured output
  const outputParser = StructuredOutputParser.fromZodSchema(
    z.object({
      actions: z.array(
        z.object({
          type: z.enum([
            'createSubtasks',
            'updateTask',
            'updateManyTasks',
            'none'
          ]),
          payload: z.any()
        })
      ),
      reasoning: z.string().describe('Your explanation of the breakdown or prioritization strategy'),
      response: z.string().describe('A concise, helpful response to the user explaining your actions and plans')
    })
  );

  // Prepare the conversation history
  const conversationHistory = state.messages.filter(msg => 
    msg.getType() === 'human' || 
    (msg.getType() === 'ai' && msg.additional_kwargs.agentType === AgentType.Planning)
  );

  // Task and user context
  const taskContext = state.task ? 
    `Task: ${state.task.title}\nDescription: ${state.task.description || 'None'}\nPriority: ${state.task.priority}\nDeadline: ${state.task.deadline ? new Date(state.task.deadline).toISOString() : 'None'}` : 
    'No task provided';

  const userContext = state.user?.psychProfile ? 
    `Productivity Time: ${state.user.psychProfile.productivityTime}\nTask Approach: ${state.user.psychProfile.taskApproach}\nDifficulty Preference: ${state.user.psychProfile.difficultyPreference}` :
    'No user profile available';

  // Create a prompt template
  const prompt = ChatPromptTemplate.fromMessages([
    ['system', getSystemPrompt('planning') + `\n\nRespond with a structured output containing actions, reasoning, and a concise user-friendly response.`],
    new MessagesPlaceholder('conversation_history'),
    ['human', `User request: {input}\n\nTask Context:\n${taskContext}\n\nUser Context:\n${userContext}\n\nProvide a structured response with actions to take in JSON format. For task breakdown, create subtasks that can be completed in 10-15 minutes each. For prioritization, consider deadlines, importance, and user preferences. Include a concise, helpful response to the user explaining your actions and plans. {format_instructions}`],
  ]);

  // Create the chain
  const chain = RunnableSequence.from([
    {
      input: (state: typeof StateAnnotation.State) => state.input,
      conversation_history: (state: typeof StateAnnotation.State) => conversationHistory,
      format_instructions: async () => outputParser.getFormatInstructions()
    },
    prompt,
    llm,
    outputParser
  ]);

  // Execute the chain
  const result = await chain.invoke(state);

  // Record the agent's thought process as a message
  if (result.reasoning) {
    state.messages.push(new AIMessage({
      content: result.reasoning,
      additional_kwargs: {
        agentType: AgentType.Planning,
        name: 'reasoning'
      }
    }));
  }

  // Store the actions for execution
  state.actionItems = result.actions.filter((action) => action.type !== 'none') as ActionItem[];
  
  // Create response for the user
  if (result.response) {
    state.agentResponse = result.response;
    
    // Push the agent's response to the message history
    state.messages.push(
      new AIMessage({
        content: state.agentResponse,
        additional_kwargs: {
          agentType: AgentType.Planning,
        },
      })
    );
  }
  
  return state;
};

File: apps/agent-python/src/agents/planning.py
from typing import Dict, Any, List
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_core.messages import AIMessage
from langchain.output_parsers import StructuredOutputParser, ResponseSchema
from ..types import State, AgentType, ActionItem
from ..utils.llm import create_llm, get_system_prompt


async def process_planning(state: State) -> Dict[str, Any]:
    """
    Process the user input with Planning agent.

    Args:
        state: Current state

    Returns:
        Dictionary of state updates
    """
    # Create LLM
    llm = create_llm('gpt-4o', 0.2)

    # Create parser for structured output
    response_schemas = [
        ResponseSchema(
            name="actions",
            description="Array of actions to take. Each action has 'type' (createSubtasks, updateTask, updateManyTasks, none) and 'payload' (any data)"
        ),
        ResponseSchema(
            name="reasoning",
            description="Your explanation of the breakdown or prioritization strategy"
        ),
        ResponseSchema(
            name="response",
            description="A concise, helpful response to the user explaining your actions and plans"
        )
    ]
    output_parser = StructuredOutputParser.from_response_schemas(response_schemas)

    # Prepare conversation history
    conversation_history = [
        msg for msg in state.messages
        if msg.type == 'human' or (
            msg.type == 'ai' and
            hasattr(msg, 'additional_kwargs') and
            msg.additional_kwargs.get('agentType') == AgentType.PLANNING.value
        )
    ]

    # Task and user context
    task_context = "No task provided"
    if state.task:
        task_context = f"""Task: {state.task.title}
Description: {state.task.description or 'None'}
Priority: {state.task.priority}
Deadline: {state.task.deadline or 'None'}"""

    user_context = "No user profile available"
    if state.user and state.user.psychProfile:
        user_context = f"""Productivity Time: {state.user.psychProfile.productivityTime or 'Not set'}
Task Approach: {state.user.psychProfile.taskApproach or 'Not set'}
Difficulty Preference: {state.user.psychProfile.difficultyPreference or 'Not set'}"""

    # Create prompt template
    prompt = ChatPromptTemplate.from_messages([
        ('system', get_system_prompt('planning') + """

Respond with a structured output containing actions, reasoning, and a concise user-friendly response."""),
        MessagesPlaceholder(variable_name='conversation_history'),
        ('human', f"""User request: {{input}}

Task Context:
{task_context}

User Context:
{user_context}

Provide a structured response with actions to take in JSON format. For task breakdown, create subtasks that can be completed in 10-15 minutes each. For prioritization, consider deadlines, importance, and user preferences. Include a concise, helpful response to the user explaining your actions and plans.

{{format_instructions}}""")
    ])

    # Create the chain
    chain = prompt | llm | output_parser

    # Execute the chain
    result = await chain.ainvoke({
        'input': state.input,
        'conversation_history': conversation_history,
        'format_instructions': output_parser.get_format_instructions()
    })

    updates = {'messages': []}

    # Record the agent's thought process as a message
    if result.get('reasoning'):
        updates['messages'].append(AIMessage(
            content=result['reasoning'],
            additional_kwargs={
                'agentType': AgentType.PLANNING.value,
                'name': 'reasoning'
            }
        ))

    # Store the actions for execution
    actions = result.get('actions', [])
    if isinstance(actions, list):
        action_items = []
        for action in actions:
            # Skip if action is a string or not a dict
            if not isinstance(action, dict):
                continue

            action_type = action.get('type')
            if action_type and action_type != 'none':
                payload = action.get('payload')
                # Ensure payload is a dict or None
                if payload is not None and not isinstance(payload, dict):
                    payload = None
                action_items.append(ActionItem(type=action_type, payload=payload))

        updates['actionItems'] = action_items

    # Create response for the user
    if result.get('response'):
        updates['agentResponse'] = result['response']
        updates['messages'].append(AIMessage(
            content=result['response'],
            additional_kwargs={'agentType': AgentType.PLANNING.value}
        ))

    return updates
